src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,236 @@
+                startNode.next = null;
+            }
+        } else if (size == 1) {
+            node = startNode;
+            startNode = null;
+        } else {
+            node.previous.next = node.next;
+            node.next.previous = node.previous;
+        }
+        size--;
+        return node.data;
+    }
+
+    private void linkLast(Node<T> node) {
+        node.previous = endNode;
+        endNode.next = node;
+        endNode = node;
+        size++;
+    }
+
+    private void linkFirst(Node<T> node) {
+        startNode.previous = node;
+        node.next = startNode;
+        startNode = node;
+        size++;
+    }
+
+    private void linkAtMiddle(Node<T> newNode, int index) {
+        Node<T> existingNode = findByIndex(index);
+        newNode.previous = existingNode.previous;
+        existingNode.previous.next = newNode;
+        newNode.next = existingNode;
+        existingNode.previous = newNode;
+        size++;
+    }
+
+    private Node<T> findByIndex(int index) {
+        indexValidation(index);
+        if (index >= 0 && index < 2 && size < 2) {
+            return index == 0 ? startNode : endNode;
+        }
+        if (index - size / 2 <= 0) {
+            return findNodeFromStart(index);
+        } else {
+            return findNodeFromEnd(index);
+        }
+    }
+
+    private Node<T> findNodeFromStart(int index) {
+        Node<T> x = startNode;
+        int currentIndex = 0;
+        while (index != currentIndex) {
+            x = x.next;
+            currentIndex++;
+        }
+        return x;
+    }
+
+    private Node<T> findNodeFromEnd(int index) {
+        Node<T> x = endNode;
+        int currentIndex = size - 1;
+        while (index != currentIndex) {
+            x = x.previous;
+            currentIndex--;
+        }
+        return x;
+    }
+
+    private void initializeBase(Node<T> node) {
+        startNode = node;
+        size++;
+    }
+
+    private void initializeTale(Node<T> node) {
+        endNode = node;
+        startNode.next = endNode;
+        endNode.previous = startNode;
+        size++;
+    }
+
+    private void indexValidation(int index) {
+        if (index < size && index >= 0) {
+            return;
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    private boolean addedWithBaseOrTaleInitialization(T value, int index) {
+        if (size == 0 && index == 0) {
+            initializeBase(new Node<>(value));
+            return true;
+        } else if (size == 1 && index == 1) {
+            initializeTale(new Node<>(value));
+            return true;
+        }
+        return false;
+    }
+
+    private boolean addedWithBaseOrTaleInitialization(T value) {
+        if (size == 0) {
+            initializeBase(new Node<>(value));
+            return true;
+        } else if (size == 1) {
+            initializeTale(new Node<>(value));
+            return true;
+        }
         return false;
     }
+
+    private class Node<T> {
```

Comment:
it can be made private static to clearly indicate its relationship with MyLinkedList